### ISSUE TITLE
Regression in 1.6.0

### ISSUE THREAD

**Issue by lovettchris** [python/mypy#16256]
**Bug Report**

The following code checks fine with mypy 1.5.1 but fails with mypy 1.6.0

**To Reproduce**

First create temp folder containing this pyproject.toml
```toml
# inform pip to use setuptools
[build-system]
requires = ["setuptools", "setuptools-scm"]
build-backend = "setuptools.build_meta"

# project metadata
[project]
name = "test_mypy_bug"
version = "0.1"
authors = [
    {name = "Test"}
]
description = "Test"
readme = "README.md"
requires-python = ">=3.9"
dependencies = [
    "pytorch-lightning==2.1.0",
    "mypy==1.6.0",
    "types-PyYAML==6.0.12.12"
]

[tool.setuptools.packages]
find = {}

[tool.mypy]
ignore_missing_imports = true
strict_optional = false

[tool.mypy-yaml]
ignore_missing_imports = true
```

Then add a test.py file containing:
```python
from pytorch_lightning import LightningModule

class PytorchLightningAgent:
    def __init__(self):
        self.model_class: type[LightningModule] = self._get_model_class()
        
    def _get_model_class(self) -> type[LightningModule]:
        raise NotImplementedError        
        
    def load_model(self, model_checkpoint_path: str) -> LightningModule:
        model = self.model_class.load_from_checkpoint(model_checkpoint_path)
        model.eval()
        return model)
```

Then run `mypy . --check` in this folder.

**Expected Behavior**

should pass because here we are actually calling a @classmethod `load_from_checkpoint`

**Actual Behavior**

```shell
(test) F:\temp\bug>mypy . --check
test.py:13: error: "object" not callable  [operator]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.6.0
- Mypy command-line flags: see above
- Mypy configuration options from `mypy.ini` (and other config files): see above
- Python version used: 3.10

This seems to be related to a recent update to the typing info in pytorch-lightning.  I wonder if they 
broke something in their type info?

**Comment by brianschubert** [comment]
Using `pytorch-lightning==2.1.0` with mypy master (1.14.0+dev.2ebc69027), I get
```none
main.py:6: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
main.py:13: error: "type[Never]" has no attribute "eval"  [attr-defined]
main.py:14: error: Incompatible return value type (got "type[Never]", expected "LightningModule")  [return-value]
Found 2 errors in 1 file (checked 1 source file)
```
If I upgrade to the latest `pytorch-lightning` (v2.4.0), I get
```none
main.py:6: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 1 source file
```
Checking `reveal_type(model)` yields `Revealed type is "pytorch_lightning.core.module.LightningModule"`, so nothing is being masked by `Any`.

It seems that if there was an issue with the typing info for pytorch-lightning, it has sense been fixed. Since there doesn't appear to be a current regression in mypy, I'm going to go ahead and close. Thanks for the original issue!
